libs/cardano-data/src/Data/OSet/Strict.hs

```diff
@@ -0,0 +1,209 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Data.OSet.Strict (
+  OSet (Empty, (:<|:), (:|>:)),
+  null,
+  size,
+  empty,
+  singleton,
+  lookup,
+  member,
+  (!?),
+  fromStrictSeq,
+  fromStrictSeqDuplicates,
+  toStrictSeq,
+  fromSet,
+  invariantHolds,
+  invariantHolds',
+  (|>),
+  (<|),
+  (|><),
+  (><|),
+)
+where
+
+import Cardano.Ledger.Binary (DecCBOR (decCBOR), EncCBOR (encCBOR))
+import Cardano.Ledger.TreeDiff (ToExpr)
+import Control.DeepSeq (NFData)
+import Data.Foldable (foldl')
+import Data.Sequence.Strict qualified as SSeq
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks)
+import Prelude hiding (lookup, null, seq)
+
+-- | A general-purpose finite, ordered, set that is strict in its
+-- values.
+--
+-- The strictness is enforced by the underlying `StrictSeq` from base,
+-- and the uniqueness of the values is enforced in this module, by the
+-- constructing functions by leveraging an accompanying `Set`.
+data OSet a = OSet
+  { osSSeq :: !(SSeq.StrictSeq a)
+  , osSet :: !(Set.Set a)
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (ToExpr, NoThunks, NFData)
+
+instance EncCBOR a => EncCBOR (OSet a) where
+  encCBOR = encCBOR . toStrictSeq
+
+instance (Show a, Ord a, DecCBOR a) => DecCBOR (OSet a) where
+  decCBOR =
+    let failOnDuplicates (duplicates, oset) =
+          if Set.null duplicates
+            then pure oset
+            else fail $ "Duplicates found while decoding OSet: " <> show duplicates
+     in failOnDuplicates . fromStrictSeqDuplicates =<< decCBOR
+
+-- | \( O(1) \). Shallow invariant using just `length` and `size`.
+invariantHolds :: OSet a -> Bool
+invariantHolds (OSet seq set) = SSeq.length seq == Set.size set
+
+-- | \( O(n \log n) \). Deep invariant using set membership check for
+-- each value. By the pigeon-hole principle, this check is exhaustive.
+invariantHolds' :: Ord a => OSet a -> Bool
+invariantHolds' oset@(OSet seq set) = invariantHolds oset && all (`Set.member` set) seq
+
+-- | \( O(1) \).
+null :: OSet a -> Bool
+null (OSet _ set) = Set.null set
+
+-- | \( O(1) \).
+size :: OSet a -> Int
+size (OSet _ set) = Set.size set
+
+-- | \( O(1) \).
+empty :: OSet a
+empty = OSet SSeq.Empty Set.empty
```

Comment:
maybe it's better to make `OSet` a `Monoid` instance instead

Reply:
There are trivially two ways to `mappend`, in the case of `OSet`, since we can have a `left-preserving` and a `right-preserving` one. And so, I don't think it is a `Monoid`, since it is not a `Semigroup`. But please correct me if I'm wrong.

Reply:
In particular `(x <> y) <> z is not trivially equal to x <> (y <> z)`. Or in other words, if I understand correctly, there are two possible `Semigroup` implementations of an `OSet`?

Reply:
What do you think @Soupstraw . Also, please correct me if I'm make any wrong statement, or any of my assumptions "smell" wrong to you :smiley:

Reply:
Alexey suggested making a `Foldable` instance too, which will require a `Monoid` instance. Do you think it is okay to pick either of the possible `mappends` as _the_ `mappend` for `Semigroup` and make `Monoid` possible?

Reply:
TODO:
1. Use `left-preserving` to match semantics of `Set` w.r.t. `Semigroup` instance. 
2. `Foldable` requires a `Monoid` instance on the type-parameter, not on the class itself.
3. Change CDDL to reflect this usage, but aliasing `OSet` to a `Set` and use the `{Enc,Dec}CBOR for `Set`, if they are order preserving.

Reply:
Yeah, both left-biased and right-biased union are valid semigroups, since they are associative.

Reply:
> Foldable requires a Monoid instance on the type-parameter, not on the class itself.

This is not the case. Replicate the Foldable instance for Sets, do not invent something new